native/ripper/ripper.go

```diff
@@ -0,0 +1,341 @@
+// The return value is a "tuple" style nested array:
+// - `Array`: contains an instruction set
+//   - `arg_types:` (none if `nil`)
+// 		 - `names:` array of names (string)
+//     - `types:` array of types (integer)
+//   - `instructions:` array of instructions
+//     - `action:` string
+//     - `anchor:` integer
+//     - `line:` integer
+//     - `params:` array of parameters (string)
+//     - `source_line:` integer
+//     - `arg_set:` (none if `nil`)
+//   		 - `names:` array of names (string)
+//       - `types:` array of types (integer)
+//
+// ```ruby
+// require 'ripper'; Ripper.instruction "10.times do |i| puts i end"
+// #=>
+//
+// require 'ripper'; Ripper.instruction "10.times do |i| puts i" # the code is invalid
+// #=> TypeError: InternalError%!(EXTRA string=String, string=Invalid Goby code)
+// ```
+//
+// @param Goby code [String]
+// @return [Array]
+func instruction(receiver Object, sourceLine int, t *Thread, args []Object) Object {
+	if len(args) != 1 {
+		return t.VM().InitErrorObject(errors.ArgumentError, sourceLine, "Expect 1 argument. got=%d", len(args))
+	}
+	
+	arg, ok := args[0].(*StringObject)
+	if !ok {
+		return t.VM().InitErrorObject(errors.TypeError, sourceLine, errors.WrongArgumentTypeFormat, classes.StringClass, arg.Class().Name)
+	}
+		
+	i, err := compiler.CompileToInstructions(arg.Value().(string), parser.NormalMode)
+	if err != nil {
+		return t.VM().InitErrorObject(errors.TypeError, sourceLine, errors.InternalError, classes.StringClass, errors.InvalidGobyCode)
+	}
+		
+	return convertToTuple(i, t.VM())
+}
+
+// Returns a nested array that contains the line #, type of the token, and the literal of the token.
+// Note that the class method does not return any errors even though the provided Goby code is invalid.
+//
+// ```ruby
+// require 'ripper'; Ripper.lex "10.times do |i| puts i end"
+// #=> [[0, "on_int", "10"], [0, "on_dot", "."], [0, "on_ident", "times"], [0, "on_do", "do"], [0, "on_bar", "|"], [0, "on_ident", "i"], [0, "on_bar", "|"], [0, "on_ident", "puts"], [0, "on_ident", "i"], [0, "on_end", "end"], [0, "on_eof", ""]]
+//
+// require 'ripper'; Ripper.lex "10.times do |i| puts i" # the code is invalid
+// #=> [[0, "on_int", "10"], [0, "on_dot", "."], [0, "on_ident", "times"], [0, "on_do", "do"], [0, "on_bar", "|"], [0, "on_ident", "i"], [0, "on_bar", "|"], [0, "on_ident", "puts"], [0, "on_ident", "i"], [0, "on_eof", ""]]
+// ```
+//
+// @param Goby code [String]
+// @return [Array]
+func lex(receiver Object, sourceLine int, t *Thread, args []Object) Object  {
+	if len(args) != 1 {
+		return t.VM().InitErrorObject(errors.ArgumentError, sourceLine, "Expect 1 argument. got=%d", len(args))
+	}
+		
+	arg, ok := args[0].(*StringObject)
+	if !ok {
+		return t.VM().InitErrorObject(errors.TypeError, sourceLine, errors.WrongArgumentTypeFormat, classes.StringClass, arg.Class().Name)
+	}
+		
+	l := lexer.New(arg.Value().(string))
+	el := t.VM().InitArrayObject([]Object{})
+	eli := []Object{}
```

Comment:
I think defining `eli` here is more for its type rather than the data itself? If so I think `var eli []Object` would be better